peps/pep-0733.rst

```diff
@@ -0,0 +1,655 @@
+it will include processes for managing the impact of the change on users
+[`Issue 60 <https://github.com/capi-workgroup/problems/issues/60>`__],
+perhaps through introducing an external backwards compatibility module
+[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
+or a new API tier of "blessed" functions
+[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].
+
+
+API Specification and Abstraction
+---------------------------------
+
+The C API does not have a formal specification, it is described
+semi-formally in the documentation and exposed through C header
+files. This creates a number of problems.
+
+Bindings for languages other than C/C++ must parse C code
+[`Issue 7 <https://github.com/capi-workgroup/problems/issues/7>`__].
+Some C language features are hard to handle in this way, because
+they produce compiler-dependent output (such as enums) or require
+a C preprocessor/compiler rather than just a parser (such as macros)
+[`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].
+
+Furthermore, C header files tend to expose more than what is intended
+to be part of the public API
+[`Issue 34 <https://github.com/capi-workgroup/problems/issues/34>`__].
+In particular, implementation details such as the precise memory
+layouts of internal data structures can be exposed
+[`Issue 22 <https://github.com/capi-workgroup/problems/issues/22>`__
+and :pep:`620`].
+This can make API evolution very difficult, in particular when it
+occurs in the stable ABI as in the case of ``ob_refcnt`` and ``ob_type``,
+which are accessed via the reference counting macros
+[`Issue 45 <https://github.com/capi-workgroup/problems/issues/45>`__].
+
+We identified a deeper issue in relation to the way that reference
+counting is exposed. The way that C extensions are required to
+manage references with calls to ``Py_INCREF`` and ``Py_DECREF`` is
+specific to CPython's memory model, and is hard for alternative
+Python implementations to emulate.
+[`Issue 12 <https://github.com/capi-workgroup/problems/issues/12>`__].
+
+Another set of problems arises from the fact that a ``PyObject*`` is
+exposed in the C API as an actual pointer rather than a handle. The
+address of an object serves as its ID and is used for comparison,
+and this complicates matters for alternative Python implementations
+that move objects during GC
+[`Issue 37 <https://github.com/capi-workgroup/problems/issues/37>`__].
+
+A separate issue is that object references are opaque to the runtime,
+discoverable only through calls to ``tp_traverse``/``tp_clear``,
+which have their own purposes. If there was a way for the runtime to
+know the structure of the object graph, and keep up with changes in it,
+this would make it possible for alternative implementations to implement
+different memory management schemes
+[`Issue 33 <https://github.com/capi-workgroup/problems/issues/33>`__].
```

Comment:
Could we add a short paragraph about defining the ABI too? Currently the ABI definition is also implicit whether having a much more explicit and simpler ABI definition would help seems an important question for future discussion.

Reply:
Can you add an issue on the problem repo that I can link to?

Reply:
The most relevant issue seems to be this discussion of the ABI -- https://github.com/capi-workgroup/problems/issues/39.

At the moment Python's "Stable ABI" is seen as some subset of the API, but really these are two completely different kinds of things that might need quite different approaches. The API is a C interface, but the ABI is compiled library binary interface.

Reply:
I don't think this issue mentions the lack of clear specification. But I think I can just add a sentence to the preamble of this section, without referring to an issue.

Reply:
The issue is more of a direct demonstration of the lack.

Reply:
> At the moment Python's "Stable ABI" is seen as some subset of the API, but really these are two completely different kinds of things that might need quite different approaches.

Is it? It's very explicitly called the Limited API and using only the limited API allows safely relying on the stable ABI.

Reply:
Yes, this is the misconception of what an ABI is that I was referring to.

Currently in CPython the C ABI is defined via the Limit API, but it doesn't have to be this way. For example, in HPy the API and ABI are rather different. The HPy API looks a lot like the old C API, but the ABI consists almost entirely of the context object (plus a few functions for determining the context version and initializing the extension).